trio/_tests/check_type_completeness.py

```diff
@@ -0,0 +1,138 @@
+#!/usr/bin/env python3
+# this file is not run as part of the tests, instead it's run standalone from check.sh
+import subprocess
+import json
+from pathlib import Path
+from typing import Any
+import sys
+import argparse
+
+RES_FILE = Path(__file__).parent / "verify_types.json"
+failed = False
+
+
+def run_pyright():
+    return subprocess.run(
+        ["pyright", "--verifytypes=trio", "--outputjson", "--ignoreexternal"],
+        capture_output=True,
+    )
+
+
+def main(args: argparse.Namespace) -> int:
+    print("*" * 20, "\nChecking type completeness hasn't gone down...")
+    current_dict: dict[str, Any] = {}
+    last_dict: dict[str, Any] = {}
+
+    def check_less_than(key, /, invert=False):
+        global failed
+        current = current_dict[key]
+        last = last_dict[key]
+        assert isinstance(current, (float, int))
+        assert isinstance(last, (float, int))
+        if current == last:
+            return
+        if (current > last) ^ invert:
+            failed = True
+            print("ERROR: ", end="")
+        if isinstance(current, float):
+            strcurrent = f"{current:.4}"
+            strlast = f"{last:.4}"
+        else:
+            strcurrent = str(current)
+            strlast = str(last)
+        print(
+            f"{key} has gone {'down' if current<last else 'up'} from {strlast} to {strcurrent}"
+        )
+
+    def check_zero(key):
+        if current_dict[key] != 0:
+            global failed
+            failed = True
+            print(f"ERROR: {key} is {current_dict[key]}")
+
+    res = run_pyright()
+    current_result = json.loads(res.stdout)
+
+    # check if py.typed file was missing
+    if (
+        current_result["generalDiagnostics"]
+        and current_result["generalDiagnostics"][0]["message"]
+        == "No py.typed file found"
+    ):
+        print("creating py.typed")
+        py_typed_file = (
+            Path(current_result["typeCompleteness"]["packageRootDirectory"])
+            / "py.typed"
+        )
+        py_typed_file.write_text("")
+
+        res = run_pyright()
+        current_result = json.loads(res.stdout)
+
+    if res.stderr:
+        print(res.stderr)
+
+    last_result = json.loads(RES_FILE.read_text())
+    current_dict = current_result["summary"]
+
+    for key in "errorCount", "warningCount", "informationCount":
+        check_zero(key)
+
+    current_dict = current_result["typeCompleteness"]
+    last_dict = last_result["typeCompleteness"]
+    for key in (
+        "missingFunctionDocStringCount",
+        "missingClassDocStringCount",
+        "missingDefaultParamCount",
+    ):
+        check_less_than(key)
+    check_less_than("completenessScore", invert=True)
+
+    current_dict = current_dict["exportedSymbolCounts"]
+    last_dict = last_dict["exportedSymbolCounts"]
```

Comment:
nit: I think it might be more obvious to index off `last_result`/`current_result` vs off `last_dict`/`current_dict`.

Reply:
yeah re-reading this now it's kinda unreadable. Should be better now